### PR TITLE
delay health checks until transport socket secrets are ready.

### PR DESCRIPTION
**Pull request by mpuncel**
This fixes a sequencing issue that could result in undesirable Envoy
bootup times. When active health checks are configured on a cluster
using a transport socket that uses SDS to fetch secrets, it was possible
for the initial health check to run before secrets are loaded, which
results in a failure. The health check would then wait for the
no_traffic_interval (default of 60s) before trying again.

This commit makes health checks wait until secrets are loaded before
starting, implemented by adding support for registering a callback on
TransportSocketFactory. SslSocketFactory supports invoking the callback
when secrets are loaded, and all other socket types currently invoke the
callback immediately, preserving existing behavior.

Signed-off-by: Michael Puncel <mpuncel@squareup.com>

Commit Message:
delay health checks until transport socket secrets are ready.

This fixes a sequencing issue that could result in undesirable Envoy
bootup times. When active health checks are configured on a cluster
using a transport socket that uses SDS to fetch secrets, it was possible
for the initial health check to run before secrets are loaded, which
results in a failure. The health check would then wait for the
no_traffic_interval (default of 60s) before trying again.

This commit makes health checks wait until secrets are loaded before
starting, implemented by adding support for registering a callback on
TransportSocketFactory. SslSocketFactory supports invoking the callback
when secrets are loaded, and all other socket types currently invoke the
callback immediately, preserving existing behavior.

Additional Description:
Risk Level: Medium
Testing: Unit tests
Docs Changes: N/A
Release Notes: Updated
Fixes https://github.com/envoyproxy/envoy/issues/12389

### CODE REVIEW

**Comment by PiotrSikora** [comment]
Drive-by comment, but wouldn't #13344 (keep clusters warming) also solve this?

**Comment by mpuncel** [comment]
@PiotrSikora interesting, I was unaware of that PR! From a quick look, I don't think that PR actually delays health checks, so it does not address the problem from https://github.com/envoyproxy/envoy/issues/12389. If there was a delay fetching a secret, Envoy would still enter the LIVE state in the middle of a `no_traffic_interval` which could be a minute long, so there could be up to a minute where Envoy is "ready" but has not performed any "real" health checks (meaning TLS secrets were loaded for it).

**Comment by mpuncel** [comment]
That PR could maybe be modified to delay `loadCluster` until secrets have loaded though, which I think would then obviate the need for this one. That might block more of the cluster initialization process than is necessary though, such as DNS resolution or EDS

**Comment by mpuncel** [comment]
Is this format check right? There is only a single line right now
 
```

check_format_test...
version_history/current.rst: Could not find 'Version history not in alphabetical order (this vs aaaa): please check placement of line' in:

ERROR: From ./version_history/current.rst
ERROR: ./version_history/current.rst:9: Version history not in alphabetical order (zzzzz vs aaaaa): please check placement of line
 * aaaaa: this should be alphabatized before z.. 
ERROR: ./version_history/current.rst:10: Version history not in alphabetical order (this should be alphabatized before z. vs aaaa is before ): please check placement of line
 * aaaaa: aaaa is before 'this'.. 
ERROR: check format failed. run 'tools/code_format/check_format.py fix'
```

File: include/envoy/network/transport_socket.h [hunk 1]
```diff
@@ -226,6 +226,13 @@ class TransportSocketFactory {
    */
   virtual TransportSocketPtr
   createTransportSocket(TransportSocketOptionsSharedPtr options) const PURE;
+
+  /**
+   * @param a callback to be invoked when the secrets required by the created transport
+   * sockets are ready. Will be invoked immediately if no secrets are required or if they
+   * are already loaded.
+   */
+  virtual void addSecretsReadyCb(std::function<void()> callback) PURE;
```
Comment (lizan) [inline include/envoy/network/transport_socket.h:235]: Name this more generic, perhaps `addReadyCb`?

**Comment by lizan** [comment]
> @PiotrSikora interesting, I was unaware of that PR! From a quick look, I don't think that PR actually delays health checks, so it does not address the problem from #12389. If there was a delay fetching a secret, Envoy would still enter the LIVE state in the middle of a `no_traffic_interval` which could be a minute long, so there could be up to a minute where Envoy is "ready" but has not performed any "real" health checks (meaning TLS secrets were loaded for it).

Envoy will still enter LIVE state but the cluster won't be active, then no HC will be there, isn't that the case?

**Comment by mpuncel** [comment]
@lizan it looks like that PR doesn't change the flow until `onClusterInit()` is called, which is after the initial round of health checks has completed, so it doesn't solve my problem.

Even after that PR is merged, I believe this problematic sequence is still possible:

1) Health checks are performed (and fail because secrets aren't loaded)
2) cluster manager properly recognizes that the cluster isn't "warmed" and waiting on secrets load
3) secrets load, Envoy goes LIVE, all clusters marked active.
4) envoy goes LIVE, requests fail with "no healthy upstream". This period could be almost 60 seconds long because of the default for `no_traffic_interval`
5) Requests succeed after the next health check interval

That said, that PR could maybe be modified to block for secrets updating _before_ calling `cluster.initialize()` or something like that, but that will also delay EDS/DNS waiting on secrets that aren't necessary.

File: source/common/upstream/health_checker_base_impl.cc [hunk 2]
```diff
@@ -384,6 +384,14 @@ void HealthCheckerImplBase::ActiveHealthCheckSession::onTimeoutBase() {
   handleFailure(envoy::data::core::v3::NETWORK);
 }
 
+void HealthCheckerImplBase::ActiveHealthCheckSession::start() {
+  // Start health checks only after secrets are ready for the transport socket
```
Comment (htuch) [inline source/common/upstream/health_checker_base_impl.cc:388]: What if you want to fetch SDS secrets from an actively health checked cluster?

Comment (mpuncel) [reply source/common/upstream/health_checker_base_impl.cc:388]: the callback is invoked in-line for all transport socket types except the TLS one. The most likely case would be that you use a raw buffer socket for the SDS cluster, health checks would start immediately in that case.

The common case would be that any SDS cluster either uses a raw buffer socket, or file based secrets which should be able to load and make progress. I could imagine a more complex setup with multiple SDS clusters that have multiple layers of secret dependencies, but at some "root" level you will need one that either doesn't use secrets or uses file based. I don't think any new challenges are created here with managing that.

Comment (htuch) [reply source/common/upstream/health_checker_base_impl.cc:388]: SG

**Review by lizan — APPROVED** [review]
Can you merge main branch?

**Comment by mpuncel** [comment]
merged!

**Review by lizan — APPROVED** [review]

**Comment by rgs1** [comment]
x-posting in case someone else bumps into this issue, we had to revert this internally because it breaks (on a significant number of cases) hot restart for us (gets stuck in PRE_INIT).

**Comment by mattklein123** [comment]
@rgs1 @mpuncel can we get this change reverted while we debug?

**Comment by rgs1** [comment]
#13639

**Comment by mpuncel** [comment]
@mattklein123 fine with me! What's the best way to discuss what might have gone wrong here and how to proceed? Should I open an issue or another PR after the revert is merged?

It sounds like the problem might be specific to hot restarting from talking with @rgs1 on slack, we didn't see the same issue in our environment where we don't use hot restart.

**Comment by mattklein123** [comment]
> @mattklein123 fine with me! Where is the best place to discuss what might have gone wrong here and how to proceed? Should I open an issue or another PR after the revert is merged?

I would reopen a PR, and then try to work with @rgs1 to debug? I don't immediately see how hot restart would change the behavior of this PR so will need some poking.

**Pull request merged**